### PR TITLE
[mqtt.homie] Add Dimmer functionality

### PR DESCRIPTION
**Pull request by szwenni**
This Pull Request adds the functionality of PercentageValue to the Homie Binding through which it is possible to link this Channel to a Dimmer Item.

This is achieved by looking at the unit value and trying to find the character `%`. If this character is found the value will be a `PercentageValue` if not the old behavior will stay.

Signed-off-by: Sven Kröger <svenni.kroeger@gmail.com>

### CODE REVIEW

File: bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java [hunk 1]
```diff
@@ -187,13 +188,16 @@ public void createChannelFromAttribute() {
...
                 BigDecimal step = (min != null && max != null)
-                        ? max.subtract(min).divide(new BigDecimal(100.0), new MathContext(isDecimal ? 2 : 0))
-                        : null;
+                           ? max.subtract(min).divide(new BigDecimal(100.0), new MathContext(isDecimal ? 2 : 0))
+                           : null;
                 if (step != null && !isDecimal && step.intValue() <= 0) {
-                    step = new BigDecimal(1);
+                        step = new BigDecimal(1);
+                }
+                if(attributes.unit != null && attributes.format.contains("%")) {
```
Comment (bodiroga) [inline bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java]: Why are you checking if the 'format' field contains the "%" sign? Shouldn't this be:

```java
if (attributes.unit != null && attributes.unit.contains("%")) {
```

Homie V4 specification defines that, for integer and float, "format" can only describe 'a range of payloads e.g. 10:15' (https://homieiot.github.io/specification/#property-attributes).

Perhaps there is something I'm misssing :thinking:

File: bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java [hunk 2]
```diff
@@ -187,13 +188,16 @@ public void createChannelFromAttribute() {
                 BigDecimal min = s.length == 2 ? convertFromString(s[0]) : null;
                 BigDecimal max = s.length == 2 ? convertFromString(s[1]) : null;
                 BigDecimal step = (min != null && max != null)
-                        ? max.subtract(min).divide(new BigDecimal(100.0), new MathContext(isDecimal ? 2 : 0))
-                        : null;
+                           ? max.subtract(min).divide(new BigDecimal(100.0), new MathContext(isDecimal ? 2 : 0))
```
Comment (bodiroga) [inline bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java]: Is there any formatting problem here? There are 3 more spaces in this lines.

File: bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java (hunk 1 above)
Comment (szwenni) [reply bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java]: I forgot to add the latest changes to this branch. I will push them as I changed this already to unit.

Not really, when I tested it OpenHab calculated the value from the format range from the Percentage value. So if your format is 0:200 and you have the Dimmer set to 50% the value which will be sent to MQTT is 100. I find it useful for values where you need dimmmer functionality but your range goes from 0 to 255 and OpenHab should display it as Dimmer.

File: bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java (hunk 2 above)
Comment (szwenni) [reply bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java]: Fixed this in the latest commit

File: bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java [hunk 3]
```diff
@@ -190,10 +191,13 @@ public void createChannelFromAttribute() {
                         ? max.subtract(min).divide(new BigDecimal(100.0), new MathContext(isDecimal ? 2 : 0))
                         : null;
                 if (step != null && !isDecimal && step.intValue() <= 0) {
-                    step = new BigDecimal(1);
+                        step = new BigDecimal(1);
```
[the remark this answers is not in the document]
Comment (szwenni) [reply bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java]: Now it looks correct, the auto formatter seems to have problems in the switch case statement somehow.

File: bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java [hunk 4]
```diff
@@ -190,10 +191,13 @@ public void createChannelFromAttribute() {
...
+                }
+                if(attributes.unit != null && attributes.unit.contains("%")) {
```
[the remark this answers is not in the document]
Comment (szwenni) [reply bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/homie300/Property.java]: Fixed it

**Comment by J-N-K** [comment]
I'm not 100% sure that this is a good idea. It will work for dimmer items, but are dimmers the only use-case for %? what about humidity values?

**Comment by bodiroga** [comment]
> I'm not 100% sure that this is a good idea. It will work for dimmer items, but are dimmers the only use-case for %? what about humidity values?

D'oh! That's, for sure, valid concern :cry: 

What about checking if the property is settable?

```java
if (attributes.unit != null && attributes.unit.contains("%") && attributes.settable) {
```

Yeah, a humidity value can also be settable, but that's a really strange use case, isn't it?

**Comment by szwenni** [comment]
I think testing if the value is settable is a good approach. If it is not settable you just want to have the numerical representation but if it is settable a percentage value can be set easier with a slider (Dimmer) than by hand in my opinion. As far as I can see there is no way to tell OpenHab that a Channel is also able capable of understanding commands of type `PercentType` that linking to a Dimmer Item is available in PaperUI.

**Comment by mjcumming** [comment]
I agree, the current implementation is confusing - if the UOM is percent then it doesn't make sense having values in the 0-1 range.  They should be 0-100.

**Comment by J-N-K** [comment]
>  As far as I can see there is no way to tell OpenHab that a Channel is also able capable of understanding commands of type `PercentType` that linking to a Dimmer Item is available in PaperUI.

If it's not a Dimmer, it should be "Number:Dimensionless" and not PercentType.

**Comment by J-N-K** [comment]
@jochen314 Can you comment on this one? I see you closed your own PR in favor of this one.

**Comment by jochen314** [comment]
I had done basically the same changes.
But this one was earlier and had more discussion, which I did not want to duplicate.

So for me, that PR looks fine, but I am not sure about the chanel type discussions here.

**Comment by cpmeister** [comment]
What is the status of this, are there still changes required?

**Comment by mjcumming** [comment]
I think before we can make this change we need some agreement with the Homie maintainers about why they are using 0-1 and a % UOM. I think most would expect a percentage value to range from 0-100 but that is not the way it is currently implemented. I am all for the change but it will break existing code my python homie module.

**Comment by bodiroga** [comment]
> I think before we can make this change we need some agreement with the Homie maintainers about why they are using 0-1 and a % UOM. I think most would expect a percentage value to range from 0-100 but that is not the way it is currently implemented. I am all for the change but it will break existing code my python homie module.

Hi @mjcumming!

Where have you read that for the "%" unit the accepted range is 0-1? I cannot find that statement here https://homieiot.github.io/specification/ :confused:

**Comment by mjcumming** [comment]
Hi,

See https://github.com/homieiot/convention/issues/190

**Comment by mjcumming** [comment]
> Hi,
> 
> See [homieiot/convention#190](https://github.com/homieiot/convention/issues/190)

The Homie maintainers haven't commented but it seems the consensus is this for percentage numbers to be interpreted as written. I.e. 100% is 100% NOT 1% as the binding current uses.

**Comment by cpmeister** [comment]
Is this PR ready to be merged then?

**Comment by mjcumming** [comment]
From my perspective due diligence has been done. I'll have to make changes to the Python [Homie](https://github.com/mjcumming/homie4) package and this will likely impact other users that have worked around the issue OH.

**Comment by cpmeister** [comment]
Manually approved sign.off

**Comment by mjcumming** [comment]
I have tested in 2.5.4 and the fix is working. Thank you.

**Comment by bodiroga** [comment]
Oh, guys, after having a look at https://github.com/openhab/openhab-addons/issues/7711 I realized that @szwenni didn't add the ```settable``` check, so the version that was merged didn't include the concerns expressed by @J-N-K. I'm sorry :disappointed: 

I will create a PR to fix it :+1:

**Pull request merged**